File: antd-cpp/examples/05-graph.cpp
#include <iostream>
#include <vector>
#include "antd/antd.hpp"

int main() {
    try {
        antd::Client client;

        // Create a graph entry (DAG node)
        auto result = client.graph_entry_put(
            "owner_secret_key_hex",
            {},  // no parents
            "content_hash_hex",
            {}   // no descendants
        );
        std::cout << "Graph entry created at: " << result.address << "\n";
        std::cout << "Cost: " << result.cost << " atto\n";

        // Read it back
        auto entry = client.graph_entry_get(result.address);
        std::cout << "Owner: " << entry.owner << "\n";
        std::cout << "Content: " << entry.content << "\n";
        std::cout << "Parents: " << entry.parents.size() << "\n";
        std::cout << "Descendants: " << entry.descendants.size() << "\n";

        // Check existence
        bool exists = client.graph_entry_exists(result.address);
        std::cout << "Exists: " << (exists ? "true" : "false") << "\n";

        // Estimate cost
        auto cost = client.graph_entry_cost("public_key_hex");
        std::cout << "Estimated cost: " << cost << " atto\n";

    } catch (const antd::AntdError& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    return 0;
}

File: antd-elixir/examples/05_graph.exs
Mix.install([
  {:antd, path: ".."}
])

# Create and query graph entries (DAG nodes)
client = Antd.Client.new()

# Create a graph entry
{:ok, result} =
  Antd.Client.graph_entry_put(client, "secret_key_hex", [], "content_hash", [
    %Antd.GraphDescendant{public_key: "pk_hex", content: "descendant_content"}
  ])

IO.puts("Graph entry created at: #{result.address}")
IO.puts("Cost: #{result.cost} atto")

# Read it back
{:ok, entry} = Antd.Client.graph_entry_get(client, result.address)
IO.puts("Owner: #{entry.owner}")
IO.puts("Content: #{entry.content}")
IO.puts("Parents: #{inspect(entry.parents)}")
IO.puts("Descendants: #{length(entry.descendants)}")

# Check existence
{:ok, exists} = Antd.Client.graph_entry_exists(client, result.address)
IO.puts("Exists: #{exists}")

# Estimate cost
{:ok, cost} = Antd.Client.graph_entry_cost(client, "public_key_hex")
IO.puts("Estimated graph entry cost: #{cost.cost} atto (#{cost.chunk_count} chunks)")
